client/src/blocking.rs

```diff
@@ -88,9 +103,74 @@ impl<'a, C: Deref<Target = impl Signer> + Clone> RequestBuilder<'a, C> {
             instruction_data: None,
             signers: Vec::new(),
             handle,
+            _phantom: PhantomData,
         }
     }
 
+    #[must_use]
+    pub fn signer<T: Signer + 'a>(mut self, signer: T) -> Self {
+        self.signers.push(Box::new(signer));
+        self
+    }
+
+    pub fn signed_transaction_with_blockhash(
+        &self,
+        latest_hash: Hash,
+    ) -> Result<Transaction, ClientError> {
+        let instructions = self.instructions()?;
+        let signers: Vec<&dyn Signer> = self.signers.iter().map(|s| s.as_ref()).collect();
+        let mut all_signers = signers;
+        all_signers.push(&*self.payer);
+
+        let tx = Transaction::new_signed_with_payer(
+            &instructions,
+            Some(&self.payer.pubkey()),
+            &all_signers,
+            latest_hash,
+        );
+
+        Ok(tx)
+    }
+
+    async fn signed_transaction_internal(&self) -> Result<Transaction, ClientError> {
+        let latest_hash =
+            AsyncRpcClient::new_with_commitment(self.cluster.to_owned(), self.options)
+                .get_latest_blockhash()
+                .await?;
+        let tx = self.signed_transaction_with_blockhash(latest_hash)?;
+
+        Ok(tx)
+    }
+
+    async fn send_internal(&self) -> Result<Signature, ClientError> {
+        let rpc_client = AsyncRpcClient::new_with_commitment(self.cluster.to_owned(), self.options);
+        let latest_hash = rpc_client.get_latest_blockhash().await?;
+        let tx = self.signed_transaction_with_blockhash(latest_hash)?;
+
+        rpc_client
+            .send_and_confirm_transaction(&tx)
+            .await
+            .map_err(Into::into)
+    }
+
+    async fn send_with_spinner_and_config_internal(
+        &self,
+        config: RpcSendTransactionConfig,
+    ) -> Result<Signature, ClientError> {
+        let rpc_client = AsyncRpcClient::new_with_commitment(self.cluster.to_owned(), self.options);
+        let latest_hash = rpc_client.get_latest_blockhash().await?;
+        let tx = self.signed_transaction_with_blockhash(latest_hash)?;
+
+        rpc_client
+            .send_and_confirm_transaction_with_spinner_and_config(
+                &tx,
+                rpc_client.commitment(),
+                config,
+            )
+            .await
+            .map_err(Into::into)
+    }
```

Comment:
> This would then not allow the compiler to be sure that all signers are threadsafe (since it would allow `Signer` and `ThreadSafeSigner`) which would put us back to square one.

Thanks for the explanation, but wouldn't the compiler be able to decide whether it's thread safe based on https://github.com/coral-xyz/anchor/blob/d8361b9df8fb80789453aa0f29d04e76339d8a88/client/src/nonblocking.rs#L28